File: bindings/kotlin/src/main/kotlin/com/aperturesyndicate/synx/SynxEngine.kt
package com.aperturesyndicate.synx

import com.aperturesyndicate.synx.internal.LibSynx
import com.sun.jna.NativeLong
import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import java.nio.charset.StandardCharsets

/**
 * Kotlin/JVM wrapper over **`synx-c`** — same engine as Rust **`synx-core` 3.6.x**
 * (`!active`, `!tool`, `.synxb`, canonical JSON). Uses **JNA** to load the native library.
 */
public object SynxEngine {
    private val lib get() = LibSynx.INSTANCE

    private fun takeString(ptr: Pointer?): String {
        if (ptr == null) throw SynxEngineError.OperationFailed
        try {
            val s = ptr.getString(0, StandardCharsets.UTF_8.name())
            return s ?: throw SynxEngineError.InvalidUtf8
        } finally {
            lib.synx_free(ptr)
        }
    }

    /** Static parse → canonical JSON string. */
    @JvmStatic
    public fun parse(text: String): String = takeString(lib.synx_parse(text))

    /** Parse with `!active` resolution → JSON string. */
    @JvmStatic
    public fun parseActive(text: String): String = takeString(lib.synx_parse_active(text))

    /** JSON (UTF-8) → SYNX text. */
    @JvmStatic
    public fun stringify(json: String): String = takeString(lib.synx_stringify(json))

    /** Canonical SYNX reformat. */
    @JvmStatic
    public fun format(text: String): String = takeString(lib.synx_format(text))

    /** `!tool` reshape → JSON string. */
    @JvmStatic
    public fun parseTool(text: String): String = takeString(lib.synx_parse_tool(text))

    /** Structural diff → JSON string. */
    @JvmStatic
    public fun diff(a: String, b: String): String = takeString(lib.synx_diff(a, b))

    /** SYNX → `.synxb` bytes. */
    @JvmStatic
    public fun compile(text: String, resolved: Boolean = false): ByteArray {
        val outLen = LongByReference()
        val p = lib.synx_compile(text, if (resolved) 1 else 0, outLen)
            ?: throw SynxEngineError.OperationFailed
        val n = outLen.value.toInt()
        return try {
            p.getByteArray(0, n)
        } finally {
            lib.synx_free_bytes(p, NativeLong(n.toLong()))
        }
    }

    /** `.synxb` → SYNX text. */
    @JvmStatic
    public fun decompile(data: ByteArray): String {
        val p = lib.synx_decompile(data, NativeLong(data.size.toLong()))
        return takeString(p)
    }

    /** True if [data] begins with `.synxb` magic. */
    @JvmStatic
    public fun isSynxb(data: ByteArray): Boolean =
        lib.synx_is_synxb(data, NativeLong(data.size.toLong())) != 0
}

File: bindings/swift/Sources/Synx/SynxEngine.swift
import CSynx
import Foundation

/// Failure from the `synx-c` FFI (`NULL` from `synx.h`).
public enum SynxEngineError: Error, Sendable {
    case operationFailed
    case invalidUTF8
}

/// Swift wrapper over **`synx-c`** — same engine as Rust **`synx-core` 3.6.x** (grammar, `!active`, `!tool`, `.synxb`, canonical JSON).
public enum SynxEngine {
    private static func takeString(_ ptr: UnsafeMutablePointer<CChar>?) throws -> String {
        guard let ptr else { throw SynxEngineError.operationFailed }
        defer { synx_free(ptr) }
        guard let s = String(validatingUTF8: ptr) else { throw SynxEngineError.invalidUTF8 }
        return s
    }

    /// Static parse → canonical JSON string.
    public static func parse(_ text: String) throws -> String {
        try text.withCString { c in
            try takeString(synx_parse(c))
        }
    }

    /// Parse with `!active` resolution → JSON string.
    public static func parseActive(_ text: String) throws -> String {
        try text.withCString { c in
            try takeString(synx_parse_active(c))
        }
    }

    /// JSON (UTF-8) → SYNX text.
    public static func stringify(json: String) throws -> String {
        try json.withCString { c in
            try takeString(synx_stringify(c))
        }
    }

    /// Canonical SYNX reformat.
    public static func format(_ text: String) throws -> String {
        try text.withCString { c in
            try takeString(synx_format(c))
        }
    }

    /// `!tool` reshape → JSON string.
    public static func parseTool(_ text: String) throws -> String {
        try text.withCString { c in
            try takeString(synx_parse_tool(c))
        }
    }

    /// Structural diff → JSON string.
    public static func diff(_ a: String, _ b: String) throws -> String {
        try a.withCString { ca in
            try b.withCString { cb in
                try takeString(synx_diff(ca, cb))
            }
        }
    }

    /// SYNX → `.synxb` bytes.
    public static func compile(_ text: String, resolved: Bool = false) throws -> Data {
        try text.withCString { c in
            var outLen: size_t = 0
            guard let p = synx_compile(c, resolved ? 1 : 0, &outLen) else {
                throw SynxEngineError.operationFailed
            }
            let count = Int(outLen)
            let data = Data(bytes: p, count: count)
            synx_free_bytes(p, outLen)
            return data
        }
    }

    /// `.synxb` → SYNX text.
    public static func decompile(_ data: Data) throws -> String {
        try data.withUnsafeBytes { raw in
            guard let base = raw.bindMemory(to: UInt8.self).baseAddress else {
                throw SynxEngineError.operationFailed
            }
            return try takeString(synx_decompile(base, data.count))
        }
    }

    /// True if data begins with `.synxb` magic.
    public static func isSynxb(_ data: Data) -> Bool {
        data.withUnsafeBytes { raw in
            guard let base = raw.bindMemory(to: UInt8.self).baseAddress else { return false }
            return synx_is_synxb(base, data.count) != 0
        }
    }
}
